nomos-services/chain/chain-leader/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+{
+    fn init(
+        service_resources_handle: OpaqueServiceResourcesHandle<Self, RuntimeServiceId>,
+        _initial_state: Self::State,
+    ) -> Result<Self, DynError> {
+        Ok(Self {
+            service_resources_handle,
+        })
+    }
+
+    #[expect(clippy::too_many_lines, reason = "TODO: Address this at some point.")]
+    async fn run(mut self) -> Result<(), DynError> {
+        let relays = CryptarchiaConsensusRelays::from_service_resources_handle::<
+            Self,
+            SamplingNetworkAdapter,
+            SamplingStorage,
+            TimeBackend,
+            CryptarchiaService,
+        >(&self.service_resources_handle)
+        .await;
+
+        // Create the API wrapper for chain service communication
+        let cryptarchia_api = CryptarchiaServiceApi::<
+            CryptarchiaService,
+            ClPool::Item,
+            RuntimeServiceId,
+        >::new::<Self>(&self.service_resources_handle)
+        .await?;
+
+        let LeaderSettings {
+            config: ledger_config,
+            transaction_selector_settings,
+            leader_config,
+            blend_broadcast_settings,
+        } = self
+            .service_resources_handle
+            .settings_handle
+            .notifier()
+            .get_updated_settings();
+
+        // TODO: check active slot coeff is exactly 1/30
+
+        let leader = Leader::new(leader_config.utxos.clone(), leader_config.sk, ledger_config);
+
+        let tx_selector = TxS::new(transaction_selector_settings);
+
+        let mut slot_timer = {
+            let (sender, receiver) = oneshot::channel();
+            relays
+                .time_relay()
+                .send(TimeServiceMessage::Subscribe { sender })
+                .await
+                .expect("Request time subscription to time service should succeed");
+            receiver.await?
+        };
+
+        let blend_adapter = BlendAdapter::<BlendService>::new(
+            relays.blend_relay().clone(),
+            blend_broadcast_settings.clone(),
+        );
+
+        self.service_resources_handle.status_updater.notify_ready();
+        info!(
+            "Service '{}' is ready.",
+            <RuntimeServiceId as AsServiceId<Self>>::SERVICE_ID
+        );
+
+        wait_until_services_are_ready!(
+            &self.service_resources_handle.overwatch_handle,
+            Some(Duration::from_secs(60)),
+            BlendService,
+            TxMempoolService<_, _, _, _, _>,
+            DaSamplingService<_, _, _, _>,
+            TimeService<_, _>
+        )
+        .await?;
+
+        let async_loop = async {
+            loop {
+                tokio::select! {
+                    Some(SlotTick { slot, .. }) = slot_timer.next() => {
+                        // TODO: Don't propose blocks until IBD is done and online mode is activated.
+                        //       Until then, mempool, DA, blend service will not be ready: https://github.com/logos-co/nomos/issues/1656
```

Comment:
Now that is separated I think we can wait for common chain to be ready, meaning that ibd is done (?)